src/core/exec-invoke.c

```diff
@@ -2440,6 +2448,91 @@ static int create_many_symlinks(const char *root, const char *source, char **sym
         return 0;
 }
 
+static int apply_exec_quotas(const char *target_dir, const char *cgroup_path, ExecDirectoryType type, uint32_t percent_quota) {
+        assert(target_dir);
+        assert(percent_quota > 0);
+        int r = 0;
+
+        dev_t devno;
+        r = get_block_device(target_dir, &devno);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to determine block device backing %s: %m", target_dir);
+        else if (devno == 0) {
+                log_warning("Not applying storage quotas for %s: File system not backed by a block device.", target_dir);
+                return 0;
+        }
+
+        /* Generate projid: cgroup path + exec directory type */
+        size_t buffer_size = strlen(cgroup_path) + snprintf(NULL, 0, "%d", INT_MIN) + 1;
+        char proj_id_plain[buffer_size];
+
+        r = snprintf(proj_id_plain, buffer_size, "%s%d", cgroup_path, type);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to generate project id for %s: %m", target_dir);
+        else if ((size_t) r >= buffer_size) {
+                log_debug("Exec quotas: Failed to generate project id for %s", target_dir);
+                return -1;
```

Comment:
please do not make up numeric error codes. we use negative errno codes, i.e. -EINVAL or so.

specifically:

```c
return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), …
```

also see coding style https://systemd.io/CODING_STYLE/#formatting